### PR TITLE
FEATURE: implement getter and setter usage for response

### PR DESCRIPTION
**Pull request by stevie-mayhew**
I had a problem on a project where something weird was happening, so I've done this.

### CODE REVIEW

**Comment by willmorgan** [comment]
`getResponse` is a good idea, but its implementation in a simple find-and-replace isn't the best here - especially if an implementor of that method performs some work every time it is called. It would make more sense to pull that out in to a variable for re-use throughout each function that uses it.

File: control/Controller.php [hunk 1]
```diff
@@ -212,9 +211,21 @@ public function getURLParams() {
...
 		return $this->response;
 	}
 
+	/**
+	 * Sets the SS_HTTPResponse object that this controller is building up.
+	 *
+	 * @param SS_HTTPResponse $response
+	 */
+	public function setResponse(SS_HTTPResponse $response) {
+		$this->response = $response;
```
Comment (dhensby) [inline control/Controller.php]: return $this

**Comment by chillu** [comment]
@stevie-mayhew On your question if we should use `$response = getResponse()` for `$this->getResponse()->addHeader(...)`, I'd say just in case there's more than one use in the same function scope - otherwise it's fine to chain.

**Comment by dhensby** [comment]
@stevie-mayhew can we get the feedback done; then we can merge :)

**Comment by stevie-mayhew** [comment]
@dhensby cool - have done.

The way which the response is handled in `handleRequest` in `LeftAndMain.php` is mighty confusing looking now this has been done.

**Comment by stevie-mayhew** [comment]
Whoops, broke the build a few times. Sorry :+1:

File: admin/code/LeftAndMain.php [hunk 2]
```diff
@@ -455,9 +457,11 @@ public function handleRequest(SS_HTTPRequest $request, DataModel $model = null)
 		if(!$response->getHeader('X-Title')) $response->addHeader('X-Title', urlencode($title));
 
 		// Prevent clickjacking, see https://developer.mozilla.org/en-US/docs/HTTP/X-Frame-Options
-		$this->response->addHeader('X-Frame-Options', 'SAMEORIGIN');
-		$this->response->addHeader('Vary', 'X-Requested-With');
-		
+		$originalResponse = $this->getResponse();
+		$originalResponse->addHeader('X-Frame-Options', 'SAMEORIGIN');
+		$originalResponse->addHeader('Vary', 'X-Requested-With');
+		$this->setResponse($originalResponse);
```
Comment (dhensby) [inline admin/code/LeftAndMain.php]: You don't need to reassign the response back as its an object so it'll be a reference, not a copy

Comment (stevie-mayhew) [inline admin/code/LeftAndMain.php]: Good point.

**Comment by sminnee** [comment]
OK it looks like @dhensby's points have been covered, I'm happy with the change, and the tests pass—merging.

File: control/Controller.php [hunk 3]
```diff
@@ -119,7 +119,6 @@ public function handleRequest(SS_HTTPRequest $request, DataModel $model) {
 		$this->pushCurrent();
 		$this->urlParams = $request->allParams();
 		$this->setRequest($request);
-		$this->response = new SS_HTTPResponse();
```
Comment (dhensby) [inline control/Controller.php]: OK - so this is a potential regression here as people are likely to require this property to be populated as part of `onBeforeInit`, `init` and `onAfterInit`

**Pull request merged**